src/core/mac/data_poll_handler.hpp

```diff
@@ -0,0 +1,231 @@
+
+        OT_STATIC_ASSERT(kMaxPollTriggeredTxAttempts < (1 << 5), "mIndirectTxAttempts cannot fit max!");
+    };
+
+    /**
+     * This class defines the callbacks used by the `DataPollHandler`.
+     *
+     */
+    class Callbacks : public InstanceLocator
+    {
+        friend class DataPollHandler;
+
+    private:
+        /**
+         * This constructor initializes the data poll handler object.
+         *
+         * @param[in]  aInstance   A reference to the OpenThread instance.
+         *
+         */
+        explicit Callbacks(Instance &aInstance);
+
+        /**
+         * This callback method requests a frame to be prepared for indirect transmission to a given sleepy child.
+         *
+         * @param[in] aFrame  A reference to a MAC frame where the new frame would be placed.
+         * @param[in] aChild  The child for which to prepare the frame.
+         *
+         * @retval OT_ERROR_NONE   Frame was prepared successfully
+         * @retval OT_ERROR_ABORT  Indirect transmission to child should be aborted (no frame for the child).
+         *
+         */
+        otError PrepareFrameForChild(Mac::Frame &aFrame, Child &aChild);
+
+        /**
+         * This callback method notifies that end of indirect frame transmission to a child.
+         *
+         * @param[in]  aFrame     The transmitted frame.
+         * @param[in]  aError     OT_ERROR_NONE when the frame was transmitted successfully,
+         *                        OT_ERROR_NO_ACK when the frame was transmitted but no ACK was received,
+         *                        OT_ERROR_CHANNEL_ACCESS_FAILURE tx failed due to activity on the channel,
+         *                        OT_ERROR_ABORT when transmission was aborted for other reasons.
+         * @param[in]  aChild     The child to which the frame was transmitted.
+         *
+         */
+        void HandleSentFrameToChild(const Mac::Frame &aFrame, otError aError, Child &aChild);
+
+        /**
+         * This callback method notifies that a pending `FrameChange` request (from `ProcessFrameChange()`) is
+         * now processed and the indirect message for the child can be updated.
+         *
+         * @param[in]  aChild     The child to update.
+         */
+        void UpdateIndirectMessage(Child &aChild);
+    };
+
+    /**
+     * This constructor initializes the data poll handler object.
+     *
+     * @param[in]  aInstance   A reference to the OpenThread instance.
+     *
+     */
+    explicit DataPollHandler(Instance &aInstance);
+
+    /**
+     * This method disables data poll handler and clear any state/info saved per child for indirect frame tx.
+     *
+     */
+    void Clear(void);
+
+    /**
+     * This method processes a frame change request (new frame, replace frame, or purge frame) for a given child.
+     *
+     * When there is an ongoing indirect frame transmission to same child the request may not be handled immediately.
```

Comment:
Rather than having the caller deal with calling `UpdateIndirectMessage()` on success, should this method simply handle that directly?

Also, the current method names do not clearly indicate the request/callback relationship (e.g. `ProcessFrameChange()` -> `UpdateIndirectMessage()`. What about having the method names share some common string that makes their relationship more obvious?

Reply:
Great idea. I updated it in the last set of pushed commits.